File: modules/checks/idor.py
from __future__ import annotations
import re
import uuid
from urllib.parse import urlparse, parse_qs, urlencode, urlunparse
import requests

from ..adaptive import Candidate
from ..findings import Finding, Severity
from ..session import session_to_curl_flags

# Patterns that look like user/resource identifiers
ID_PATTERN = re.compile(r'(\d{1,10})')
UUID_PATTERN = re.compile(
    r'[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}',
    re.IGNORECASE,
)
PATH_ID_PATTERN = re.compile(r'/([\w-]+)/(\d{1,10})(?:/|$)')


def run(session: requests.Session, base_url: str, endpoints: list, store) -> None:
    print("[*] Checking IDOR/BOLA...")
    curl_auth = session_to_curl_flags(session)
    timeout = getattr(session, 'timeout', 15)

    for ep in endpoints:
        if ep.method not in ("GET", "HEAD"):
            continue
        _check_numeric_ids_in_path(session, ep, curl_auth, timeout, store)
        _check_numeric_ids_in_params(session, ep, curl_auth, timeout, store)
        _check_uuid_in_path(session, ep, curl_auth, timeout, store)


def _check_numeric_ids_in_path(session, ep, curl_auth, timeout, store):
    parsed = urlparse(ep.url)
    path = parsed.path

    # Find segments like /users/123 or /api/v1/accounts/456
    for match in PATH_ID_PATTERN.finditer(path):
        resource = match.group(1)
        current_id = int(match.group(2))

        probes = _adjacent_ids(current_id)

        original_resp = _get(session, ep.url, timeout)
        if original_resp is None:
            continue
        original_len = len(original_resp.text)
        original_code = original_resp.status_code

        if original_code not in (200, 201):
            continue

        for probe_id in probes:
            new_path = path[:match.start(2)] + str(probe_id) + path[match.end(2):]
            new_url = urlunparse((
                parsed.scheme, parsed.netloc, new_path,
                parsed.params, parsed.query, ""
            ))
            resp = _get(session, new_url, timeout)
            if resp is None:
                continue

            if resp.status_code == 200 and abs(len(resp.text) - original_len) < len(resp.text) * 0.5:
                if len(resp.text) > 50:
                    store.add_candidate(Candidate(
                        finding=Finding(
                            title=f"Потенциальный IDOR — /{resource}/{{id}} доступен без смены владельца",
                            severity=Severity.MEDIUM,
                            category="IDOR / BOLA",
                            cwe="CWE-639",
                            description=(
                                f"Запрос к /{resource}/{probe_id} вернул 200 и ответ, по объёму схожий "
                                f"с оригиналом. Текущий ID: {current_id}, проверенный: {probe_id}. "
                                "Если пользователь может читать данные другого пользователя — это BOLA "
                                "(Broken Object Level Authorization). Похожий объём ответа сам по себе "
                                "не доказывает доступ к чужим данным — требуется проверка содержимого."
                            ),
                            url=new_url,
                            evidence=(
                                f"Оригинал ({current_id}): HTTP {original_code}, {original_len} байт\n"
                                f"Зонд ({probe_id}): HTTP {resp.status_code}, {len(resp.text)} байт\n"
                                f"Фрагмент ответа: {resp.text[:200]}"
                            ),
                            remediation=(
                                "1. Проверяйте на стороне сервера, что объект принадлежит аутентифицированному пользователю.\n"
                                "2. Используйте indirect references (UUID) вместо последовательных числовых ID.\n"
                                "3. Централизованная авторизация на уровне ORM/репозитория."
                            ),
                            reproduction=(
                                f"# Оригинальный запрос:\n"
                                f"curl -sk {curl_auth} '{ep.url}'\n\n"
                                f"# Замена ID на соседний:\n"
                                f"curl -sk {curl_auth} '{new_url}'"
                            ),
                        ),
                        kind="idor",
                        context={"original_resp": original_resp, "probe_resp": resp},
                    ))
                    break  # One candidate per endpoint is enough


def _check_numeric_ids_in_params(session, ep, curl_auth, timeout, store):
    parsed = urlparse(ep.url)
    qs = parse_qs(parsed.query)

    id_params = {k: v[0] for k, v in qs.items()
                 if v and ID_PATTERN.fullmatch(v[0]) and k.lower() in (
                     "id", "user_id", "userid", "account_id", "accountid",
                     "customer_id", "customerid", "uid", "pid", "oid",
                     "order_id", "orderid", "profile_id", "doc_id", "file_id",
                 )}

    if not id_params:
        return

    original_resp = _get(session, ep.url, timeout)
    if original_resp is None or original_resp.status_code not in (200, 201):
        return
    original_len = len(original_resp.text)

    for param, current_val in id_params.items():
        current_id = int(current_val)
        for probe_id in _adjacent_ids(current_id):
            new_qs = qs.copy()
            new_qs[param] = [str(probe_id)]
            new_url = urlunparse((
                parsed.scheme, parsed.netloc, parsed.path,
                parsed.params, urlencode(new_qs, doseq=True), ""
            ))
            resp = _get(session, new_url, timeout)
            if resp is None:
                continue
            if resp.status_code == 200 and len(resp.text) > 50:
                if abs(len(resp.text) - original_len) < len(resp.text) * 0.6:
                    store.add(Finding(
                        title=f"Потенциальный IDOR — параметр '{param}'",
                        severity=Severity.HIGH,
                        category="IDOR / BOLA",
                        cwe="CWE-639",
                        description=(
                            f"Параметр '{param}={probe_id}' (заменён с {current_id}) "
                            "вернул успешный ответ. Возможна утечка данных другого пользователя."
                        ),
                        url=ep.url,
                        parameter=param,
                        evidence=(
                            f"Оригинал ({current_id}): {original_len} байт\n"
                            f"Зонд ({probe_id}): {len(resp.text)} байт\n"
                            f"Ответ: {resp.text[:200]}"
                        ),
                        remediation=(
                            "1. Авторизация на уровне объекта — проверьте owner_id == current_user.id.\n"
                            "2. Используйте GUID/UUID вместо числовых ID.\n"
                            "3. Логируйте запросы с чужими ID."
                        ),
                        reproduction=(
                            f"# Оригинал:\ncurl -sk {curl_auth} '{ep.url}'\n\n"
                            f"# IDOR:\ncurl -sk {curl_auth} '{new_url}'"
                        ),
                    ))
                    break


def _check_uuid_in_path(session, ep, curl_auth, timeout, store):
    parsed = urlparse(ep.url)
    uuids = UUID_PATTERN.findall(parsed.path)
    if not uuids:
        return

    original_resp = _get(session, ep.url, timeout)
    if original_resp is None or original_resp.status_code not in (200, 201):
        return

    for current_uuid in uuids:
        probe_uuid = str(uuid.uuid4())
        new_path = parsed.path.replace(current_uuid, probe_uuid)
        new_url = urlunparse((
            parsed.scheme, parsed.netloc, new_path,
            parsed.params, parsed.query, ""
        ))
        resp = _get(session, new_url, timeout)
        if resp is None:
            continue
        if resp.status_code == 200 and len(resp.text) > 100:
            store.add(Finding(
                title="Потенциальный IDOR — замена UUID в пути вернула 200",
                severity=Severity.MEDIUM,
                category="IDOR / BOLA",
                cwe="CWE-639",
                description=(
                    f"Замена UUID '{current_uuid}' на случайный '{probe_uuid}' "
                    "вернула HTTP 200. Требует ручной проверки — возможна авторизация без проверки владельца."
                ),
                url=ep.url,
                evidence=f"Новый UUID: {probe_uuid}\nСтатус: {resp.status_code}\nОтвет: {resp.text[:200]}",
                remediation=(
                    "Убедитесь, что проверяется принадлежность объекта аутентифицированному пользователю."
                ),
                reproduction=(
                    f"# Оригинал:\ncurl -sk {curl_auth} '{ep.url}'\n\n"
                    f"# Замена UUID:\ncurl -sk {curl_auth} '{new_url}'"
                ),
            ))


def _adjacent_ids(current: int) -> list[int]:
    probes = []
    for delta in (1, 2, -1, -2):
        candidate = current + delta
        if candidate > 0:
            probes.append(candidate)
    return probes[:4]


def _get(session, url, timeout):
    try:
        return session.get(url, timeout=timeout)
    except Exception:
        return None

File: go-vectrix/internal/checks/idor.go
package checks

import (
	"crypto/rand"
	"fmt"
	"net/url"
	"regexp"
	"strconv"
	"strings"

	"vectrixgo/internal/crawler"
	"vectrixgo/internal/findings"
	"vectrixgo/internal/httpsession"
	"vectrixgo/internal/logging"
)

var (
	idPattern     = regexp.MustCompile(`^\d{1,10}$`)
	uuidPattern   = regexp.MustCompile(`(?i)[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`)
	pathIDPattern = regexp.MustCompile(`/([\w-]+)/(\d{1,10})(?:/|$)`)
)

var idorParamNames = map[string]struct{}{
	"id": {}, "user_id": {}, "userid": {}, "account_id": {}, "accountid": {},
	"customer_id": {}, "customerid": {}, "uid": {}, "pid": {}, "oid": {},
	"order_id": {}, "orderid": {}, "profile_id": {}, "doc_id": {}, "file_id": {},
}

// RunIDOR checks GET/HEAD endpoints for sequential numeric IDs / UUIDs in the
// path or query string that can be swapped for adjacent values. Mirrors
// modules/checks/idor.py run().
func RunIDOR(session *httpsession.Session, baseURL string, endpoints []crawler.Endpoint, store *findings.FindingStore) {
	logging.Println("[*] Checking IDOR/BOLA...")
	curlAuth := session.CurlAuthFlags(baseURL)

	for _, ep := range endpoints {
		if ep.Method != "GET" && ep.Method != "HEAD" {
			continue
		}
		checkNumericIDsInPath(session, ep, curlAuth, store)
		checkNumericIDsInParams(session, ep, curlAuth, store)
		checkUUIDInPath(session, ep, curlAuth, store)
	}
}

func checkNumericIDsInPath(session *httpsession.Session, ep crawler.Endpoint, curlAuth string, store *findings.FindingStore) {
	parsed, err := url.Parse(ep.URL)
	if err != nil {
		return
	}
	path := parsed.Path

	for _, m := range pathIDPattern.FindAllStringSubmatchIndex(path, -1) {
		resource := path[m[2]:m[3]]
		currentID, err := strconv.Atoi(path[m[4]:m[5]])
		if err != nil {
			continue
		}

		probes := adjacentIDs(currentID)

		originalResp, err := getResp(session, ep.URL)
		if err != nil {
			continue
		}
		originalLen := len(originalResp.Body)
		originalCode := originalResp.StatusCode
		if originalCode != 200 && originalCode != 201 {
			continue
		}

		for _, probeID := range probes {
			newPath := path[:m[4]] + strconv.Itoa(probeID) + path[m[5]:]
			newU := *parsed
			newU.Path = newPath
			newU.Fragment = ""
			newURL := newU.String()

			resp, err := getResp(session, newURL)
			if err != nil {
				continue
			}

			if resp.StatusCode == 200 && absInt(len(resp.Body)-originalLen) < int(float64(len(resp.Body))*0.5) {
				if len(resp.Body) > 50 {
					f := findings.NewFinding(
						fmt.Sprintf("Потенциальный IDOR — /%s/{id} доступен без смены владельца", resource),
						findings.Medium,
						"IDOR / BOLA",
						"CWE-639",
						newURL,
						fmt.Sprintf("Запрос к /%s/%d вернул 200 и ответ, по объёму схожий "+
							"с оригиналом. Текущий ID: %d, проверенный: %d. "+
							"Если пользователь может читать данные другого пользователя — это BOLA "+
							"(Broken Object Level Authorization). Похожий объём ответа сам по себе "+
							"не доказывает доступ к чужим данным — требуется проверка содержимого.",
							resource, probeID, currentID, probeID),
						"1. Проверяйте на стороне сервера, что объект принадлежит аутентифицированному пользователю.\n"+
							"2. Используйте indirect references (UUID) вместо последовательных числовых ID.\n"+
							"3. Централизованная авторизация на уровне ORM/репозитория.",
						fmt.Sprintf("# Оригинальный запрос:\ncurl -sk %s '%s'\n\n"+
							"# Замена ID на соседний:\ncurl -sk %s '%s'", curlAuth, ep.URL, curlAuth, newURL),
					)
					f.Evidence = fmt.Sprintf("Оригинал (%d): HTTP %d, %d байт\n"+
						"Зонд (%d): HTTP %d, %d байт\n"+
						"Фрагмент ответа: %s",
						currentID, originalCode, originalLen,
						probeID, resp.StatusCode, len(resp.Body), truncate(resp.Body, 200))

					store.AddCandidate(&findings.Candidate{
						Finding: f,
						Kind:    "idor",
						Context: map[string]any{"original_resp": originalResp, "probe_resp": resp},
					})
					break // One candidate per endpoint is enough
				}
			}
		}
	}
}

func checkNumericIDsInParams(session *httpsession.Session, ep crawler.Endpoint, curlAuth string, store *findings.FindingStore) {
	parsed, err := url.Parse(ep.URL)
	if err != nil {
		return
	}
	qs := parsed.Query()

	idParams := make(map[string]string)
	for k, v := range qs {
		if len(v) == 0 {
			continue
		}
		if !idPattern.MatchString(v[0]) {
			continue
		}
		if _, ok := idorParamNames[strings.ToLower(k)]; ok {
			idParams[k] = v[0]
		}
	}
	if len(idParams) == 0 {
		return
	}

	originalResp, err := getResp(session, ep.URL)
	if err != nil || (originalResp.StatusCode != 200 && originalResp.StatusCode != 201) {
		return
	}
	originalLen := len(originalResp.Body)

	for param, currentVal := range idParams {
		currentID, err := strconv.Atoi(currentVal)
		if err != nil {
			continue
		}
		for _, probeID := range adjacentIDs(currentID) {
			newQS := cloneValues(qs)
			newQS.Set(param, strconv.Itoa(probeID))
			newU := *parsed
			newU.RawQuery = newQS.Encode()
			newU.Fragment = ""
			newURL := newU.String()

			resp, err := getResp(session, newURL)
			if err != nil {
				continue
			}
			if resp.StatusCode == 200 && len(resp.Body) > 50 {
				if absInt(len(resp.Body)-originalLen) < int(float64(len(resp.Body))*0.6) {
					f := findings.NewFinding(
						fmt.Sprintf("Потенциальный IDOR — параметр '%s'", param),
						findings.High,
						"IDOR / BOLA",
						"CWE-639",
						ep.URL,
						fmt.Sprintf("Параметр '%s=%d' (заменён с %d) "+
							"вернул успешный ответ. Возможна утечка данных другого пользователя.", param, probeID, currentID),
						"1. Авторизация на уровне объекта — проверьте owner_id == current_user.id.\n"+
							"2. Используйте GUID/UUID вместо числовых ID.\n"+
							"3. Логируйте запросы с чужими ID.",
						fmt.Sprintf("# Оригинал:\ncurl -sk %s '%s'\n\n# IDOR:\ncurl -sk %s '%s'", curlAuth, ep.URL, curlAuth, newURL),
					)
					f.Parameter = param
					f.Evidence = fmt.Sprintf("Оригинал (%d): %d байт\nЗонд (%d): %d байт\nОтвет: %s",
						currentID, originalLen, probeID, len(resp.Body), truncate(resp.Body, 200))
					store.Add(f)
					break
				}
			}
		}
	}
}

func checkUUIDInPath(session *httpsession.Session, ep crawler.Endpoint, curlAuth string, store *findings.FindingStore) {
	parsed, err := url.Parse(ep.URL)
	if err != nil {
		return
	}
	uuids := uuidPattern.FindAllString(parsed.Path, -1)
	if len(uuids) == 0 {
		return
	}

	originalResp, err := getResp(session, ep.URL)
	if err != nil || (originalResp.StatusCode != 200 && originalResp.StatusCode != 201) {
		return
	}

	for _, currentUUID := range uuids {
		probeUUID := randomUUID()
		newPath := strings.Replace(parsed.Path, currentUUID, probeUUID, 1)
		newU := *parsed
		newU.Path = newPath
		newU.Fragment = ""
		newURL := newU.String()

		resp, err := getResp(session, newURL)
		if err != nil {
			continue
		}
		if resp.StatusCode == 200 && len(resp.Body) > 100 {
			f := findings.NewFinding(
				"Потенциальный IDOR — замена UUID в пути вернула 200",
				findings.Medium,
				"IDOR / BOLA",
				"CWE-639",
				ep.URL,
				fmt.Sprintf("Замена UUID '%s' на случайный '%s' "+
					"вернула HTTP 200. Требует ручной проверки — возможна авторизация без проверки владельца.",
					currentUUID, probeUUID),
				"Убедитесь, что проверяется принадлежность объекта аутентифицированному пользователю.",
				fmt.Sprintf("# Оригинал:\ncurl -sk %s '%s'\n\n# Замена UUID:\ncurl -sk %s '%s'", curlAuth, ep.URL, curlAuth, newURL),
			)
			f.Evidence = fmt.Sprintf("Новый UUID: %s\nСтатус: %d\nОтвет: %s", probeUUID, resp.StatusCode, truncate(resp.Body, 200))
			store.Add(f)
		}
	}
}

func adjacentIDs(current int) []int {
	var probes []int
	for _, delta := range []int{1, 2, -1, -2} {
		candidate := current + delta
		if candidate > 0 {
			probes = append(probes, candidate)
		}
	}
	if len(probes) > 4 {
		probes = probes[:4]
	}
	return probes
}

func getResp(session *httpsession.Session, rawurl string) (*httpsession.Response, error) {
	return session.Get(rawurl, nil)
}

func absInt(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func cloneValues(v url.Values) url.Values {
	out := make(url.Values, len(v))
	for k, vals := range v {
		cp := make([]string, len(vals))
		copy(cp, vals)
		out[k] = cp
	}
	return out
}

func randomUUID() string {
	b := make([]byte, 16)
	rand.Read(b)
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
